### PR TITLE
Pipeline Caching CacheBeta 1.0 agent changes

### PR DESCRIPTION
**Pull request by johnterickson**
Addresses:
- Fallback keys: microsoft/azure-pipelines-tasks#10842
- input globbing: microsoft/azure-pipelines-tasks#10859
- Relative paths: microsoft/azure-pipelines-tasks#10858
- Only save when restore runs: microsoft/azure-pipelines-tasks#10907
- Proxy

TODO:
[✔]Waiting on new package
[ punted ] Attributes/TAR: microsoft/azure-pipelines-tasks#10841

### CODE REVIEW

**Comment by johnterickson** [comment]
supercedes https://github.com/microsoft/azure-pipelines-agent/pull/2358

**Comment by johnterickson** [comment]
@TingluoHuang This is ready to merge.  It's also needed to be able to take new packages from master.

**Comment by TingluoHuang** [comment]
@johnterickson is anyone from your side reviewed it? :D

**Comment by johnterickson** [comment]
Yup! This is just a rebase of https://github.com/microsoft/azure-pipelines-agent/pull/2340 :)

@owenhuynMSFT and @fadnavistanmay to explicitly sign off on this one too

File: src/Agent.Plugins/PipelineCache/RestorePipelineCacheV0.cs [hunk 1]
```diff
@@ -10,23 +12,20 @@ public class RestorePipelineCacheV0 : PipelineCacheTaskPluginBase
         public override string Stage => "main";
 
         protected override async Task ProcessCommandInternalAsync(
-            AgentTaskPluginExecutionContext context, 
-            string path, 
-            string keyStr,
-            string salt,
+            AgentTaskPluginExecutionContext context,
```
Comment (fadnavistanmay) [inline src/Agent.Plugins/PipelineCache/RestorePipelineCacheV0.cs]: Can we create a new version for this change?

File: src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs [hunk 2]
```diff
@@ -1,61 +1,135 @@
...
+                    segments = (new [] { $"{SaltVariableName}={salt}"}).Concat(segments);
+                }
+                return segments.ToArray();
+            };
+
+            Func<string,string[]> splitAcrossNewlines = (s) => 
+                s.Replace("\r\n", "\n") //normalize newlines
+                 .Split(new [] {'\n'}, StringSplitOptions.RemoveEmptyEntries)
+                 .Select(line => line.Trim())
+                 .ToArray();
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs]: ParseIntoSegements is being used in the old format and the new format. I thought that the V0 plugin won't support the new format while the v1 plugin will support new and old.

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            }
+
+            foreach (string keySegment in keySegments)
+            {
+                if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToArray();
+                    string[] includeRules = pathRules.Where(p => !p.StartsWith('!')).ToArray();
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: >'!' [](start = 79, length = 3)

What does '!' mean in terms of a path segment?

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+using System.Text;
+
+[assembly: InternalsVisibleTo("Test")]
+
+namespace Agent.Plugins.PipelineCache
+{
+    public static class FingerprintCreator
+    {
+        private static readonly bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+        private static readonly bool isCaseSensitive = RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
```
Comment (fadnavistanmay) [inline src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: Add mac as well?

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        private static bool IsPathyChar(char c)
+        {
+            if (GlobChars.Contains(c)) return true;
+            if (c == Path.DirectorySeparatorChar) return true;
+            if (c == Path.AltDirectorySeparatorChar) return true;
+            if (c == Path.VolumeSeparatorChar) return true;
+            return !Path.GetInvalidFileNameChars().Contains(c);
+        }
+
+        internal static bool IsPathy(string keySegment)
```
Comment (fadnavistanmay) [inline src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: IsPathyKeySegment?

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs (hunk 4 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: For?

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs (hunk 3 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: "except" - same as gitignore and artifactignore https://docs.microsoft.com/en-us/azure/devops/artifacts/reference/artifactignore?view=azure-devops

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                        List<string> globs;
+                        if(!enumerations.TryGetValue(enumeration, out globs))
+                        {
+                            enumerations[enumeration] = globs = new List<string>(); 
+                        }
+                        globs.Add(absoluteRootRule);
+                    }
+
+                    string[] excludeRules = pathRules.Where(p => p.StartsWith('!')).ToArray();
+                    string[] absoluteExcludeRules = excludeRules.Select(excludeRule => {
```
Comment (owenhuynMSFT) [inline src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: I'd say we leverage off a public dotnet globbing library:
[https://github.com/dazinator/DotNet.Glob](https://github.com/dazinator/DotNet.Glob)

**Comment by owenhuynMSFT** [comment]
In the description this PR I don't think address these but other PRs have been merged to address them:
Only save when restore runs: microsoft/azure-pipelines-tasks#10907
Retries on timeouts: microsoft/azure-pipelines-tasks#10935

File: src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs [hunk 7]
```diff
@@ -1,61 +1,135 @@
...
+        private const string SaltVariableName = "AZDEVOPS_PIPELINECACHE_SALT";
+        private const string OldKeyFormatMessage = "'key' format is changing to a single line: https://aka.ms/pipeline-caching-docs";
+
         public Guid Id => PipelineCachePluginConstants.CacheTaskId;
 
         public abstract String Stage { get; }
 
+        internal static (bool isOldFormat, string[] keySegments,IEnumerable<string[]> restoreKeys) ParseIntoSegments(string salt, string key, string restoreKeysBlock)
+        {
+            Func<string,string[]> splitIntoSegments = (s) => {
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs]: >splitIntoSegments [](start = 34, length = 17)

splitAcrossPipes

File: src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs [hunk 8]
```diff
@@ -135,7 +114,25 @@ public class PipelineCacheServer
...
+                            context.Verbose($"This fingerprint: `{fingerprint.ToString()}`");
+
+                            if(fingerprint == result.Fingerprint)
+                            {
+                                foundExact = true;
+                                break;
+                            }
+                        }
+
+                        context.SetVariable(cacheHitVariable, foundExact ? "true" : "inexact");
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs]: >inexact [](start = 85, length = 7)

Wording?
Guidance should be to either eq/ne on true/false (i.e. don't eq/ne on inexact)

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+            if (!keySegment.Contains(".") && 
+                !keySegment.Contains(Path.DirectorySeparatorChar) &&
+                !keySegment.Contains(Path.AltDirectorySeparatorChar)) return false;
+            if (keySegment.Last() == '.') return false;
+            return true;
+        }
+
+        internal static Func<string,bool> CreateMinimatchFilter(AgentTaskPluginExecutionContext context, string rule, bool invert)
+        {
+            Func<string,bool> filter = Minimatcher.CreateFilter(rule, minimatchOptions);
```
Comment (johnterickson) [inline src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: >Minimatcher [](start = 39, length = 11)

this vs. DotNet.Glob

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs (hunk 4 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: OSX is case-insensitive like Windows

---
In reply to: [306966039](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306966039) [](ancestors = 306966039)

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs (hunk 3 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: Does Minimatch already handle the commas?

---
In reply to: [306966358](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306966358) [](ancestors = 306966358)

File: src/Agent.Plugins/PipelineCache/FingerprintCreator.cs (hunk 6 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/FingerprintCreator.cs]: or do we match the rest of DevOps? 🤔 @willsmythe?

---
In reply to: [306972347](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306972347) [](ancestors = 306972347)

File: src/Agent.Plugins/PipelineCache/RestorePipelineCacheV0.cs (hunk 1 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/RestorePipelineCacheV0.cs]: I think we decided against - is that right?

---
In reply to: [306960674](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306960674) [](ancestors = 306960674)

File: src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs (hunk 2 above)
Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs]: I think Will and I landed on the v0 task supporting both (warning on old) and then the v1 task will only support the new format.

**Comment by fadnavistanmay** [comment]
@TingluoHuang - let's merge this. :) When are you planning to start agent 155 btw? Ton of fixes in that version :D

File: src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs [hunk 10]
```diff
@@ -1,61 +1,135 @@
...
 
 namespace Agent.Plugins.PipelineCache
 {
     public abstract class PipelineCacheTaskPluginBase : IAgentTaskPlugin
     {
+        protected const string RestoreStepRanVariableName = "RESTORE_STEP_RAN";
+        protected const string RestoreStepRanVariableValue = "true";
+
+        private const string SaltVariableName = "AZDEVOPS_PIPELINECACHE_SALT";
+        private const string OldKeyFormatMessage = "'key' format is changing to a single line: https://aka.ms/pipeline-caching-docs";
```
Comment (tiwilson7) [inline src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs]: When will the documentation be updated? I had to dig up this file to figure out what the new key format is. The linked documentation still shows the old multi-line format.

Comment (johnterickson) [reply src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs]: @willsmythe  Will you be handling the new docs?

**Pull request merged**